src/incremental/compareAST.ml

```diff
@@ -3,212 +3,284 @@ open CilMaps
-    | TNamed (tinf, attr), b -> eq_typ_acc tinf.ttype b ~rename_mapping ~acc  (* Ignore tname, treferenced. TODO: dismiss attributes, or not? *)
-    | a, TNamed (tinf, attr) -> eq_typ_acc a tinf.ttype ~rename_mapping ~acc  (* Ignore tname, treferenced . TODO: dismiss attributes, or not? *)
+  let r, updated_rename_mapping = match a, b with
+    | TPtr (typ1, attr1), TPtr (typ2, attr2) -> eq_typ_acc typ1 typ2 ~rename_mapping ~acc &&>> forward_list_equal (eq_attribute ~acc) attr1 attr2
+    | TArray (typ1, (Some lenExp1), attr1), TArray (typ2, (Some lenExp2), attr2) -> eq_typ_acc typ1 typ2 ~rename_mapping ~acc &&>> eq_exp lenExp1 lenExp2 ~acc &&>>  forward_list_equal (eq_attribute ~acc) attr1 attr2
+    | TArray (typ1, None, attr1), TArray (typ2, None, attr2) -> eq_typ_acc typ1 typ2 ~rename_mapping ~acc &&>> forward_list_equal (eq_attribute ~acc) attr1 attr2
+    | TFun (typ1, (Some list1), varArg1, attr1), TFun (typ2, (Some list2), varArg2, attr2) ->
+      eq_typ_acc typ1 typ2 ~rename_mapping ~acc &&>>
+      forward_list_equal (eq_args ~fun_parameter_name_comparison_enabled:fun_parameter_name_comparison_enabled ~acc) list1 list2 &&>
+      (varArg1 = varArg2) &&>>
+      forward_list_equal (eq_attribute ~acc) attr1 attr2
+    | TFun (typ1, None, varArg1, attr1), TFun (typ2, None, varArg2, attr2) ->
+      eq_typ_acc typ1 typ2 ~rename_mapping ~acc &&>
+      (varArg1 = varArg2) &&>>
+      forward_list_equal (eq_attribute ~acc) attr1 attr2
+    | TNamed (typinfo1, attr1), TNamed (typeinfo2, attr2) ->
+      eq_typ_acc typinfo1.ttype typeinfo2.ttype ~rename_mapping ~acc &&>> forward_list_equal (eq_attribute ~acc) attr1 attr2 (* Ignore tname, treferenced *)
+    | TNamed (tinf, attr), b -> eq_typ_acc tinf.ttype b ~rename_mapping ~acc (* Ignore tname, treferenced. TODO: dismiss attributes, or not? *)
+    | a, TNamed (tinf, attr) -> eq_typ_acc a tinf.ttype ~rename_mapping ~acc (* Ignore tname, treferenced . TODO: dismiss attributes, or not? *)
     (* The following two lines are a hack to ensure that anonymous types get the same name and thus, the same typsig *)
     | TComp (compinfo1, attr1), TComp (compinfo2, attr2) ->
       if mem_typ_acc a b acc || mem_typ_acc a b !global_typ_acc then (
         if Messages.tracing then Messages.trace "compareast" "in acc\n";
-        true
+        true, rename_mapping
       )
       else (
         let acc = (a, b) :: acc in
-        let res = eq_compinfo compinfo1 compinfo2 acc rename_mapping && GobList.equal (eq_attribute ~rename_mapping ~acc) attr1 attr2 in
-        if res then begin
-          global_typ_acc := (a, b) :: !global_typ_acc;
-
-          (* Reset cnames and ckeys to the old value. Only affects anonymous structs/unions where names are not checked for equality. *)
-          compinfo2.cname <- compinfo1.cname;
-          compinfo2.ckey <- compinfo1.ckey;
```

Comment:
They do. They are executed in `performRenames` after a successful comparison. It was pulled out such that the comparison itself has no side effects in case an assumed match did not pass the comparison and is discarded again.
https://github.com/goblint/analyzer/blob/6ba71fc9a44d593a76b259ac2b6b0a6a209cc899/src/incremental/compareCIL.ml#L90-L95

Reply:
Right, makes sense. Would the old comment about anonymous structs be still relevant (and should be copied over here) or not?

Reply:
I think it is still helpful and so I copied it.